### ISSUE TITLE
SqlSetup fails to run after configuring ForcedEncryption in SqlSecureConnection

### ISSUE THREAD

**Issue by Kreby** [dsccommunity/SqlServerDsc#1893]
### Problem description

The ```SqlSetup``` resource fails to execute on subsequent runs after an initial install if a ```SqlSecureConnection``` resource has been processed and has  ```ForceEncryption = $true```.  The issue is related to the lack of support for FQDN's, much like in #1888.

This specifics of this issue is two fold 

Firstly the ```ServerName``` property for ```SqlSetup``` [is assumed](https://github.com/dsccommunity/SqlServerDsc/blob/4c98dda0f50a9eb564ff8e949d4f1cb0cfa05325/source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.psm1#L159) and obtained using the ```Get-ComputerName``` function.  This only becomes a problem after Encryption is forced.

Secondly, [the root of the first issue](https://github.com/dsccommunity/DscResource.Common/blob/bdcb5330000a192591f6dff630fd97915326349a/source/Public/Get-ComputerName.ps1#L33),  is the function ```Get-ComputerName``` from the ```DscResource.Common``` module will only ever provide a hostname and is not capable of returning a FQDN.

The ```ServerName``` variable is then passed to a number of other cmdlets which in turn call ```Connect-Sql``` and we've seen before in, #1888,  the use of a FQDN is required in order to connect.

### Verbose logs

```text
Connect-Sql : System.InvalidOperationException: Failed to connect to SQL instance 'localhost'. (SQLCOMMON0019) --->
System.Management.Automation.MethodInvocationException: Exception calling "Connect" with "0" argument(s): "Failed to
connect to server localhost." ---> Microsoft.SqlServer.Management.Common.ConnectionFailureException: Failed to connect
to server localhost. ---> Microsoft.Data.SqlClient.SqlException: A connection was successfully established with the
server, but then an error occurred during the login process. (provider: SSL Provider, error: 0 - The target principal
name is incorrect.) ---> System.ComponentModel.Win32Exception: The target principal name is incorrect
   --- End of inner exception stack trace ---
```

### DSC configuration

```powershell
It's not really relevant to the issue but can provide if needed.
```

### Suggested solution

The best option would probably be to add a ```ServerName``` property to the ```SqlSetup``` resource which defaults to just using the ```Get-ComputerName``` it is already assuming when the argument isn't supplied.  

If adding a ```ServerName``` property is not acceptable then trying to resolve the hostname via DNS is an option but I feel that might be more problematic.  I too was using ```$env:COMPUTERNAME``` previously but currently I am working around that by relying on DNS

```powershell
$serverName = [System.Net.Dns]::GetHostByName($env:COMPUTERNAME).HostName 
```

I do plan to go back to fix to pass the FQDN around but I haven't done that work yet.

I am currently only testing the DNS lookup on Server 2022 so I can't speak to it's support on other OS's as well older versions of Windows or PowerShell at the moment.  I've not looked up to see what versions of Dot Net use this. 

### SQL Server edition and version

```text
Microsoft SQL Server 2022 (RTM) - 16.0.1000.6 (X64)   Oct  8 2022 05:58:25   Copyright (C) 2022 Microsoft Corporation  Standard Edition (64-bit) on Windows Server 2022 Datacenter 10.0 <X64> (Build 20348: ) (Hypervisor)
```

### SQL Server PowerShell modules

```text
Name            Version    Path
----            -------    ----
SqlServer       22.0.49    C:\Program Files\WindowsPowerShell\Modules\SqlServer\22.0.49\SqlServer.psd1
SqlServer       21.1.18256 C:\Program Files\WindowsPowerShell\Modules\SqlServer\21.1.18256\SqlServer.psd1
SqlServer       21.1.18221 C:\Program Files\WindowsPowerShell\Modules\SqlServer\21.1.18221\SqlServer.psd1
SQLPS           16.0       C:\Program Files (x86)\Microsoft SQL Server\160\Tools\PowerShell\Modules\SQLPS\SQLPS.psd1
```

### Operating system

```text
OsName               : Microsoft Windows Server 2022 Datacenter
OsOperatingSystemSKU : DatacenterServerEdition
OsArchitecture       : 64-bit
WindowsVersion       : 2009
WindowsBuildLabEx    : 20348.1.amd64fre.fe_release.210507-1500
OsLanguage           : en-US
OsMuiLanguages       : {en-US}
```

### PowerShell version

```text
Name                           Value
----                           -----
PSVersion                      5.1.20348.1366
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.20348.1366
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```

### SqlServerDsc version

```text
Name         Version Path
----         ------- ----
SqlServerDsc 16.1.0  C:\Program Files\WindowsPowerShell\Modules\SqlServerDsc\16.1.0\SqlServerDsc.psd1
```

**Comment by johlju** [comment]
This get the information on current node on macOS so it might work on all OS'es, although it doesn't return FQDN there, but it is not connected to an AD domain either, so not a good example. 🙂 

```
[System.Net.Dns]::GetHostByName($null)
```

I think `Get-ComputerName` returns what is expected, it could possible be extended with a switch parameter `FQDN` that runs the above command to return the fully qualified domain name . 🤔  That would help solve most (or all) resource by adding that parameter to `Get-ComputerName`.

But it would not help if the network name to connect to is different than the hostname. For example when using AG or failover clustering. For failover clustering there is the parameter `FailoverClusterNetworkName`, but after creation the [dns name is gotten from the cluster](https://github.com/dsccommunity/SqlServerDsc/blob/4c98dda0f50a9eb564ff8e949d4f1cb0cfa05325/source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.psm1#L3001). But I wonder if we still need a `ServerName` parameter to handle AG's or other scenrios where network name is different from hostname. 🤔

**Comment by johlju** [comment]
Seems `GetHostByName()` is deprecated, instead we should use this to return the entry for the current node:

```powershell
[System.Net.Dns]::GetHostEntry('')
```

**Comment by Kreby** [comment]
I'm not currently running an AG in the configuration I'm testing with and that is something I hadn't yet thought about.  

I did consider proposing a FQDN switch for the ```Get-ComputerName```.  That might certainly be helpful in general, but then how do you decide when to use the switch in the DSC resources?  I'm not sure if you want to default to using it or not.  I imagine that could have a number of unexpected outcomes for people who expect/want just the hostname?

I guess I should switch over to using ```GetHostEntry()``` on my side of things even if it is temporary.  

I did consider that the DNS lookup could fail.  I didn't have a better option when I started testing.  This would also require the SQL server to have a network connection and a reachable DNS server (assuming that's how ```GetHostEntry``` actually works).  It is a bit naive to think that will always be the case, if I'm being honest.  So I don't know if there is a better way than to add the ```ServerName``` property to the resource and let the configuration author decide what needs to be passed.

**Comment by johlju** [comment]
@Kreby I made a suggested fix in PR https://github.com/dsccommunity/SqlServerDsc/pull/1900 using ServerName. Could you verify that it works?

**Comment by johlju** [comment]
@Kreby I merged the suggested fix. If there are any issue or if it doesn't work out as expected let me know and we make a change before next full release is released.

**Comment by Kreby** [comment]
@johlju Thanks for the turn around on that.   The newly released  ```SqlServer``` module has caused me a bit of a headache elsewhere so I've had to focus my efforts on addressing some other issues first.  I hope to be able to test this tomorrow.  I'll let you know how it goes.

**Comment by Kreby** [comment]
@johlju I was able to test it this morning and verify that everything was working as expected now. Subsequent runs are no longer failing.  I did it before you released the latest ```16.2.0``` version so my tests were just using the main branch.

Thanks again for the help.

**Comment by johlju** [comment]
Yes I went ahead and released the new version so we would get the support for SqlServer 22.0.59. Happy it worked as expected though, thanks for testing and reporting back. Then we squashed that one too. 😃
